examples/gcf-pubsub-vm-delete-event-handler/main.py

```diff
@@ -0,0 +1,660 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+__author__ = ('Jeff McCune <jeff@openinfrastructure.co>, '
+              'Gary Larizza <gary@openinfrastructure.co>')
+
+import base64
+import json
+import logging
+import os
+import google.cloud.logging
+from enum import Enum
+from google.cloud.logging.resource import Resource
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from typing import List
+
+
+class RuntimeState:
+    """Stores App instance for the lifetime of the process"""
+    pass
+
+
+RuntimeState.app = None
+
+
+class Result(Enum):
+    """The overall result of the DNS cleanup for user reporting
+
+    An OK result indicates the cleanup completed normally.
+
+    A NOT_PROCESSED results is likely a result of losing the race against the
+    VM delete operation and is intended to signal to the user they may need to
+    cleanup DNS records using another mechanism (e.g. manually).
+    """
+    OK = 0
+    NOT_PROCESSED = 1
+
+
+class Severity(Enum):
+    """Stackdriver severity levels
+
+    See https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+    """
+    DEFAULT = 0
+    DEBUG = 100
+    INFO = 200
+    NOTICE = 300
+    WARNING = 400
+    ERROR = 500
+    CRITICAL = 600
+    ALERT = 700
+    EMERGENCY = 800
+
+
+class Detail(Enum):
+    """Detailed results of the cleanup
+
+    A LOST_RACE result indicates user intervention is necessary.
+    """
+    NO_OP = 0
+    NO_MATCHES = 1
+    RR_DELETED = 2
+    VM_NO_IP = 3
+    IGNORED_EVENT = 4
+    LOST_RACE = 5
+    RR_MISMATCH = 6
+    RR_NOT_A_RECORD = 7
+    RR_NAME_MISMATCH = 8
+    RR_IP_MISMATCH = 9
+
+
+class StructuredLog(object):
+    """Base class to report structured log events
+
+    Sub classes are expected to override MESSAGE (str), RESULT (Result), and
+    DETAIL (Detail).
+
+    """
+    MESSAGE = '{} mark'
+    RESULT = None
+    DETAIL = None
+    SEVERITY = Severity.NOTICE
+    LEVEL = logging.INFO
+
+    def __init__(self, vm_uri: str, event_id: str):
+        self.vm_uri = vm_uri
+        self.event_id = event_id
+        self.function_project = os.getenv('GCP_PROJECT')
+        self.function_region = os.getenv('FUNCTION_REGION')
+        self.function_name = os.getenv('FUNCTION_NAME')
+        stream = os.getenv('DNS_VM_GC_REPORTING_LOG_STREAM')
+        # The log stream structured reports are sent to.
+        # https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+        if stream:
+            self.log_name = stream
+        else:
+            self.log_name = (
+                'projects/{}/logs/{}'
+            ).format(self.function_project, self.function_name)
```

Comment:
Nit: use f string for brevity / readability.
GCF run python 3.7
As of python 3.6 we have formatted string literals [PEP 498](https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep498)
```python
self.log_name = f"projects/{self.function_project}/{self.function_name}"
```

Cosider using this instead of `str.format()` throughout this file.

Reply:
Ack, will do.  Thanks for letting me know, it's been ~10 years since I last wrote serious Python.

Reply:
Sure thing! New features are hard to keep track of even when you write python every day!